piq/fsim.py

```diff
@@ -140,24 +135,33 @@ def _construct_filters(x: torch.Tensor, scales: int = 4, orientations: int = 4,
         delta_theta: Ratio of angular interval between filter orientations
             and the standard deviation of the angular Gaussian function
             used to construct filters in the freq. plane.
-        k: No of standard deviations of the noise energy beyond the mean
-            at which we set the noise threshold point, below which phase
-            congruency values get penalized.
-        """
+    Returns:
+          Tensor with filters. Shape :math:`(1, scales * orientations, H, W)`
+    """
     N, _, H, W = x.shape
 
     # Calculate the standard deviation of the angular Gaussian function
     # used to construct filters in the freq. plane.
     theta_sigma = math.pi / (orientations * delta_theta)
 
     # Pre-compute some stuff to speed up filter construction
-    grid_x, grid_y = get_meshgrid((H, W))
+    grid_x, grid_y = get_meshgrid((H, W), device=x.device, dtype=x.dtype)
 
-    # Move grid to GPU early on, so that all math heavy stuff computes faster.
-    grid_x, grid_y = grid_x.to(x), grid_y.to(x)
     radius = torch.sqrt(grid_x ** 2 + grid_y ** 2)
     theta = torch.atan2(-grid_y, grid_x)
 
+    # First construct a low-pass filter that is as large as possible, yet falls
+    # away to zero at the boundaries.  All log Gabor filters are multiplied by
+    # this to ensure no extra frequencies at the 'corners' of the FFT are
+    # incorporated as this seems to upset the normalisation process when
+    # Computed explicitly without _lowpassfilter
+    # Explicit low pass filter computation
+    n = 15  # default parameter
+    cutoff = .45  # default parameter
+    assert 0 < cutoff <= 0.5, "Cutoff frequency must be between 0 and 0.5"
+    assert n > 1 and int(n) == n, "n must be an integer >= 1"
```

Comment:
Could you elaborate why these assets are necessary considering that `cutoff` and `n` values are not parameters but hardcoded constants?

Reply:
The hardcoded values and secretion check came from the previous implementation.
While the common use-case would be not bothering with these constants making assertions redundant, some edge cases of enthusiast wanting to vary this parameter could be tracked better.
While it is not a huge overhead in load and the same computation path as before, we do not induce additional functionality keeping an advanced user safe.